spec/services/theme_settings_migrations_runner_spec.rb

```diff
@@ -0,0 +1,285 @@
+
+      expect do described_class.new(theme).run end.to raise_error(
+        Theme::SettingsMigrationError,
+        I18n.t("themes.import_error.migrations.invalid_filename", filename: "020-some-name"),
+      )
+
+      migration_field.update_attribute(:name, "0020some-name")
+
+      expect do described_class.new(theme).run end.to raise_error(
+        Theme::SettingsMigrationError,
+        I18n.t("themes.import_error.migrations.invalid_filename", filename: "0020some-name"),
+      )
+
+      migration_field.update_attribute(:name, "0020")
+
+      expect do described_class.new(theme).run end.to raise_error(
+        Theme::SettingsMigrationError,
+        I18n.t("themes.import_error.migrations.invalid_filename", filename: "0020"),
+      )
+    end
+
+    it "raises an error if a pending migration has version lower than the last ran migration" do
+      migration_field.update!(name: "0020-some-name")
+      Fabricate(:theme_settings_migration, theme: theme, theme_field: migration_field, version: 20)
+
+      Fabricate(:migration_theme_field, theme: theme, version: 19, name: "0019-failing-migration")
+
+      expect do described_class.new(theme).run end.to raise_error(
+        Theme::SettingsMigrationError,
+        I18n.t(
+          "themes.import_error.migrations.out_of_sequence",
+          name: "0019-failing-migration",
+          current: 20,
+        ),
+      )
+    end
+
+    it "detects bad syntax in migrations and raises an error" do
+      migration_field.update!(value: <<~JS)
+        export default function migrate() {
+      JS
+      expect do described_class.new(theme).run end.to raise_error(
+        Theme::SettingsMigrationError,
+        I18n.t(
+          "themes.import_error.migrations.syntax_error",
+          name: "0001-some-name",
+          error:
+            'SyntaxError: "/discourse/theme/migration: Unexpected token (2:0)\n\n  1 | export default function migrate() {\n> 2 |\n    | ^"',
+        ),
+      )
+    end
+
+    it "imposes memory limit on migrations and raises an error if they exceed the limit" do
+      migration_field.update!(value: <<~JS)
+        export default function migrate(settings) {
+          let a = new Array(10000);
+          while(true) {
+            a = a.concat(new Array(10000));
+          }
+          return settings;
+        }
+      JS
+
+      expect do described_class.new(theme, memory: 10.kilobytes).run end.to raise_error(
+        Theme::SettingsMigrationError,
+        I18n.t("themes.import_error.migrations.exceeded_memory_limit", name: "0001-some-name"),
+      )
+    end
+
+    it "imposes time limit on migrations and raises an error if they exceed the limit" do
+      migration_field.update!(value: <<~JS)
+        export default function migrate(settings) {
+          let a = 1;
+          while(true) {
+            a += 1;
+          }
+          return settings;
+        }
+      JS
+
+      expect do described_class.new(theme, timeout: 10).run end.to raise_error(
```

Comment:
I think we need a better way to test this since this is going to increase the test suite by 10 seconds.

Reply:
oh the unit here is milliseconds, not seconds 😄

Reply:
Oooo icic then this is ok :grin: